File: mir/mir.cpp
/*
 * Copyright (C) <year> <name of author>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include <iostream>

char const *next;    

bool S(); 
bool A(); 
bool A1(); 
bool A2();

bool term(char token) {
    if (*next != '\0')
        return *next++ == token;
    else
        return false;
}

bool S() {
    return term('c') && A() && term('d');
}   

    
bool A() {
    char const *save = next;
    return A1() || (next = save, A2());
}

bool A1() {
    return term('a') && term('b');
}
bool A2() {
    return term('a') ;
}

int main() {
    next = "caed";
    
    if (S() && *next == '\0') {
        std::cout << "aceptada";
    }
    else
        std::cout << "no aceptada";
}


/*
 * Editor modelines  -  https://www.wireshark.org/tools/modelines.html
 *
 * Local variables:
 * c-basic-offset: 8
 * tab-width: 8
 * indent-tabs-mode: t
 * End:
 *
 * vi: set shiftwidth=8 tabstop=8 noexpandtab:
 * :indentSize=8:tabSize=8:noTabs=false:
 */

File: mir/Mir.java
/*
 * Copyright (C) <year> <name of author>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

public class Mir {
    private String next;
    private int index;

    public Mir(String _next) {
        next = _next;
        index = 0;
    }

    private boolean term(char token) {
        if ( index != next.length() )
            return next.charAt(index++) == token;
        else
            return false;
    }

    private boolean S() {
        return term('c') && A() && term('d');
    }

    private boolean A() {
        int save = index;
        //return A1() || (index = save, A2());
        if ( ! A1() ) {
            index=save;
            return A2();
        }
        return true;
    }

    private boolean A1() {
        return term('a') && term('b');
    }

    private boolean A2() {
        return term('a');
    }

    public boolean test() {
        return S() && index == next.length();
    }

    public static void main(String args[]) {
        Mir m = new Mir("caed");

        if ( m.test() ) {
            System.out.println("aceptada");
        } else {
            System.out.println("no aceptada");
        }
    }
}

/*
 * Editor modelines  -  https://www.wireshark.org/tools/modelines.html
 *
 * Local variables:
 * c-basic-offset: 4
 * tab-width: 4
 * indent-tabs-mode: nil
 * End:
 *
 * vi: set shiftwidth=4 tabstop=4 expandtab:
 * :indentSize=4:tabSize=4:noTabs=true:
 */

